store/src/java/com/zimbra/cs/html/owasp/OwaspHtmlSanitizer.java

```diff
@@ -77,12 +83,27 @@ public void handle(final String x) {
         instantiatePolicy();
         final Policy policy = POLICY_DEFINITION.apply(new StyleTagReceiver(renderer));
         // run the html through the sanitizer
-        HtmlSanitizer.sanitize(html, policy);
+        HtmlSanitizer.sanitize(cleanMalformedHtml(html), policy);
         // return the resulting HTML from the builder
         OwaspHtmlSanitizer.zThreadLocal.remove();
         return htmlBuilder.toString();
     }
 
+    private String cleanMalformedHtml(String str) throws UnsupportedEncodingException {
+        Tidy tidy = new Tidy();
+        tidy.setInputEncoding("UTF-8");
+        tidy.setOutputEncoding("UTF-8");
+        tidy.setPrintBodyOnly(true);
+        tidy.setXHTML(true);
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(str.getBytes("UTF-8"));
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        tidy.parseDOM(inputStream, outputStream);
+        if (outputStream.toString("UTF-8").isEmpty()) {
+            return str;
+        }
+        return outputStream.toString("UTF-8");
```

Comment:
Did you test multibyte non-utf8 characters?

Reply:
Will check and update.

Reply:
**Testing with multibyte non-utf8 characters e-mails:**
<img width="1289" alt="EUC-JP" src="https://user-images.githubusercontent.com/3608604/113309353-b03fa980-9324-11eb-8a35-b8e3315868c6.png">

<img width="1254" alt="Shift_JIS" src="https://user-images.githubusercontent.com/3608604/113309489-d2d1c280-9324-11eb-9aba-57c8eaff2454.png">

-------------

**Malformed multibyte non-utf8 characters e-mails:**
**Without using JTidy:**
<img width="594" alt="Multi-Byte_Malformed_Japanese_Without_JTidy" src="https://user-images.githubusercontent.com/3608604/113309673-01e83400-9325-11eb-8b12-35f3954bb98e.png">

-------------

**Using JTidy:**
<img width="599" alt="Multi-Byte_Malformed_Japanese_With_JTidy" src="https://user-images.githubusercontent.com/3608604/113309727-11677d00-9325-11eb-96a4-cbd9cec803ab.png">

-------------

**Malformed JIS e-mails:**
<img width="1252" alt="Malformed JIS e-mails" src="https://user-images.githubusercontent.com/3608604/113310146-8935a780-9325-11eb-8bdb-321db1099326.png">